release process draft

edublancas opened the pull request.

idomic commented on release-process.md:
```diff
@@ -0,0 +1,106 @@
+
+These are changes that we add to the `CHANGELOG`:
+
+- [Doc] Important documentation changes
```
Those should rarely get in imo. It will probably also affect the way we use the CD.

idomic replied:
unless we map on prefix, still didn't get there.

edublancas replied:
> unless we map on prefix, still didn't get there.

I'm not following

idomic replied:
We can rule out docs changes based on the [Doc] prefix. I think if only important ones get into the changelog it's fine and it should work fine.

idomic commented on release-process.md:
```diff
@@ -0,0 +1,106 @@
+- [Doc] Important documentation changes
+- [Feature] New features
+- [API Change] API breaking changes
+- [Fix] Bug fixes
```
Is this the common convention?

edublancas replied:
yeah, loosely based on this: https://scikit-learn.org/stable/whats_new/v1.2.html

idomic commented on release-process.md:
```diff
@@ -0,0 +1,106 @@
+what happens next. Most often, the following action would be to deploy a new minor version
+to PyPI, so once the PR is merged and the CI passes in `master`/`main`, a new minor version will be uploaded to PyPI.
+
+Another action can be "no release", implying that the PR is merged, but no release
```
How does the reviewer pass these flags?

idomic replied:
Also, shouldn't we make it automatic?
This adds complexity to the review process which is already pretty time consuming.

edublancas replied:
> How does the reviewer pass these flags?

to the commit message when merging

> Also, shouldn't we make it automatic?

This is challenging to automate. How can a script determine if this is a minor or major release? It'd have to analyze the API and look for breaking changes. This also adds accountability since the reviewer is responsible for signing off.

edublancas replied:
just a clarifying note: I'm not against having some automation here. but I think this has to be a human decision. however, we can implement it with software. e.g., write a program that checks for API breaking changes so the reviewer can quickly determine if this will be a major release (opened an issue here https://github.com/ploomber/pkgmt/issues/30); however, I think we shouldn't wait to have that program ready to get started with this

idomic replied:
Sounds good ^^
As long as it's not blocking I'm fine. I do think for the major/minor you're right, maybe this script can be a check in the CI so the reviewer knows there's a warning about backward compatibility.

edublancas replied:
ok, let's leave the issue open on pkgmt  and for now, the reviewer must sign off the release

idomic commented on release-process.md:
```diff
@@ -0,0 +1,106 @@
+so we should carefully plan with deprecation warnings to give our users
+a heads-up. In this case, once the PR is merged, a new major release is created (via `git tag`), and pushed to PyPI.
+
+The reviewer will indicate what action to take by adding one *command* when merging
```
Maybe move to the beginning of the section

idomic commented on release-process.md:
````diff
@@ -0,0 +1,106 @@
+`master`/`main` branch passes. We can tag a new release with the following command:
+
+```sh
+invoke version
````
This is the manual process that we currently have it's just to doc it? I thought this is the part we're automating.

edublancas replied:
check the note at the beginning of the section.

idomic reviewed (commented):
Added comments

yafimvo reviewed (commented):
Since many tests fail due to lint issues, mentioning our lint tools here might help.

edublancas commented:
> Since many tests fail due to lint issues, mentioning our lint tools here might help.

good point. I'll add this

Note: I just bumped into an edge we hadn't considered. we also need to check that the docs build correctly, otherwise if the docs break, we might end up releasing something without releasing the docs at the same time

idomic commented:
> Note: I just bumped into an edge we hadn't considered. we also need to check that the docs build correctly, otherwise if the docs break, we might end up releasing something without releasing the docs at the same time

That's a good point, probably we want to add a dependency on the docs build to the CI

edublancas commented:
> Since many tests fail due to lint issues, mentioning our lint tools here might help.

added a short section on this.

> That's a good point, probably we want to add a dependency on the docs build to the CI

Yest, this is a quick but functional fix. I think it'll work 95% of the time. 5% it might be that the docs pass but readthedocs has some error and doesn't deploy. but it'll do for now.

Added some note on this.

@yafimvo @idomic alright thanks for your feedback. I think we'r in good shape for v1. Once this is merged, I can create an issue to track the progress of implementing the CD script.
